### PR TITLE
Tests fix

### PR DESCRIPTION
**Pull request by baktrius**
- Extend `test` script with option to specify which tests to run allowing creation of bundles of tests (with directories)
- Move all test from `tests/` directory to `tests/basic` to create space for students tests
- Fix all tests to not depend on `mimpirun` return code propagation

### CODE REVIEW

File: examples/test.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#define SUCCESS_MARKER "<<success>>\n"
+
+int test_success()
+{
+    // below unchecked syscall return code
+    write(STDERR_FILENO, SUCCESS_MARKER, sizeof(SUCCESS_MARKER));
+    return 0;
+}
+
+#define test_assert(expr) assert(expr && "<<error>>")
```
Comment (wprzytula) [inline examples/test.h:15]: I don't undestand how it is supposed to work. Why _logical and_ over `expr` and a string literal??

File: run_test [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+tmp_file=$(mktemp)
+
+# Invoke mimpirun with appropriate timeout and arguments
+# "|| true" used to ignore ./mimpirun return code
+timeout "$TIMEOUT" ./mimpirun "$NUM_OF_WORKERS" "$@" 2> "$tmp_file" || true
+
+result=0
+
+# Check if all workers finished sucessfully
+# - num of <<success>> is equeal to number of created workers
```
Comment (wprzytula) [inline run_test:26]: equal*

File: run_test [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+# Check if error has been detected - <<error>> was found in stderr
+grep "<<error>>" "$tmp_file" > /dev/null && result=2
+
+# Print stderr in case of error
+test "$result" -ne 0 && cat "$tmp_file" >&2
+
+# Remove temporary file
+rm "$tmp_file"
+
+# Exit with appropraite return code
```
Comment (wprzytula) [inline run_test:38]: appropriate

**Review by wprzytula — CHANGES_REQUESTED** [review]

File: examples/test.h (hunk 1 above)
Comment (baktrius) [reply examples/test.h:15]: It's based on the following post:
https://stackoverflow.com/questions/3692954/add-custom-messages-in-assert

**Pull request merged**